torch/_dynamo/variables/builder.py

```diff
@@ -443,11 +443,12 @@ def index_source(key):
                 value, guards=make_guards(GuardBuilder.FUNCTION_MATCH)
             )
         elif (
-            isinstance(value, types.BuiltinFunctionType)
-            and type(getattr(value, "__self__", None))
+            type(getattr(value, "__self__", None))
             is torch.autograd.function.FunctionMeta
             and getattr(value, "__name__", "") == "apply"
             and value == getattr(value.__self__, "apply", None)
+            and hasattr(value, "__func__")
+            and value.__func__ is torch.autograd.Function.apply.__func__
```

Comment:
Note on dynamo changes:
- this PR modifies autograd.Function's apply method. It is no longer a "BuiltinFunctionType" (i.e. it is no longer a C++ function).
- AFAICT dynamo here checks if we have an autograd.Function that does not have a user-overiden apply method.
- So I modified this to check that the apply function is indeed not overiden